### PR TITLE
Msp refactor

### PR DESCRIPTION
**Pull request by makarandp0**
We have lot of boilerplate code to handle MSP channels with VMS failovers, This change refactors it to remove boilerplate. 
This was previously opened as: https://github.com/twilio/twilio-video.js/pull/1386
**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

**Comment by makarandp0** [comment]
Note: the failing test is caused by [some SFU change](https://twilio.slack.com/archives/C7J8VNN3W/p1613679305089800?thread_ts=1613676373.088600&cid=C7J8VNN3W) in stage:

File: lib/signaling/v2/msptransporthandler.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        writable: true
+      }
+    });
+  }
+
+  isSetup() {
+    return !!this._receiverPromise;
+  }
+
+  isReady() {
```
[the remark this answers is not in the document]
Comment (manjeshbhargav) [reply lib/signaling/v2/msptransporthandler.js:42]: Also, `isReady()` is not used anywhere, so you can remove it.

**Review by manjeshbhargav — CHANGES_REQUESTED** [review]
@makarandp0 Getting better! However, I have a suggestion. Right now, we seem to have two types of classes: `MediaSignaling` (formerly `MSPTransportHandler`) classes for each MSP channel and the different media signaling classes: `DominantSpeakerSignaling`, `TrackSwitchOffSignaling`, etc. I think it makes sense for each of these media signaling classes to derive from `MediaSignaling`. For ex:

```js
class DominantSpeakerSignaling extends MediaSignaling {
  ...
}
```
Now, instead of having 2 objects `_dominantSpeakerMSP` and `dominantSpeakerSignaling`, we can have only one object `dominantSpeakerSignaling`. We can now listen to both "ready" and "updated" events on the same object.

**Comment by makarandp0** [comment]
Thanks @manjeshbhargav. I like your suggestion of collapsing MSP Transport with MSP Signaling, however I am afraid that would be a larger change than I wanted to take up - especially on the unit test side - we have lot of unit tests for this signaling classes and such change would invalidate them.

How about we postpone extended refactor for some other day. With the current refactor we have already reduces the code significantly, and made it easy to add new MSP as shown by https://github.com/twilio/twilio-video.js/pull/1391/files

File: lib/signaling/v2/dominantspeakersignaling.js [hunk 2]
```diff
@@ -21,14 +20,16 @@ class DominantSpeakerSignaling extends EventEmitter {
...
 
-    mediaSignalingTransport.on('message', message => {
-      switch (message.type) {
-        case 'active_speaker':
-          this._setLoudestParticipantSid(message.participant);
-          break;
-        default:
-          break;
-      }
+    this.on('ready', mediaSignalingTransport => {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/dominantspeakersignaling.js:23]: Nitpick:
```suggestion
    this.on('ready', transport => {
```

File: lib/signaling/v2/networkqualitysignaling.js [hunk 3]
```diff
@@ -110,14 +106,16 @@ class NetworkQualitySignaling extends EventEmitter {
...
 
-    mediaSignalingTransport.on('message', message => {
-      switch (message.type) {
-        case 'network_quality':
-          this._handleNetworkQualityMessage(message);
-          break;
-        default:
-          break;
-      }
+    this.on('ready', mediaSignalingTransport => {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/networkqualitysignaling.js:109]: Nitpick:
```suggestion
    this.on('ready', transport => {
```

File: lib/signaling/v2/trackprioritysignaling.js [hunk 4]
```diff
@@ -1,15 +1,30 @@
...
     Object.defineProperties(this, {
-      _mediaSignalingTransport: {
-        value: mediaSignalingTransport
-      }
+      _enqueuedPriorityUpdates: {
+        value: new Map()
+      },
+    });
+
+    this.on('ready', mediaSignalingTransport => {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/trackprioritysignaling.js:18]: NitpicK:
```suggestion
    this.on('ready', transport => {
```

File: lib/signaling/v2/trackswitchoffsignaling.js [hunk 5]
```diff
@@ -1,25 +1,27 @@
...
-      switch (message.type) {
-        case 'track_switch_off':
-          this._setTrackSwitchOffUpdates(message.off || [], message.on || []);
-          break;
-        default:
-          break;
-      }
+  constructor(getReceiver, options) {
+    super(getReceiver, 'track_switch_off', options);
+    this.on('ready', mediaSignalingTransport => {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/trackswitchoffsignaling.js:15]: Nitpick:
```suggestion
    this.on('ready', transport => {
```

File: test/unit/spec/signaling/v2/room.js [hunk 6]
```diff
@@ -1409,9 +1281,9 @@ describe('RoomV2', () => {
               sinon.assert.callCount(dataTrackReceiver2.toDataTransport, 1);
             });
 
-            it('should construct a TrackPrioritySignaling with the new DataTrackTransport,', () => {
-              sinon.assert.calledWith(TrackPrioritySignaling, dataTrackTransport2);
-            });
+            // it('should construct a TrackPrioritySignaling with the new DataTrackTransport,', () => {
+            //   sinon.assert.calledWith(TrackPrioritySignaling, dataTrackTransport2);
+            // });
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/room.js:1286]:
```suggestion
```

File: test/unit/spec/signaling/v2/room.js [hunk 7]
```diff
@@ -1579,9 +1439,9 @@ describe('RoomV2', () => {
           assert(dataTrackReceiver.toDataTransport.calledOnce);
         });
 
-        it('constructs a DominantSpeakerSignaling with the DataTrackTransport,', () => {
-          assert(DominantSpeakerSignaling.calledWith(dataTrackTransport));
-        });
+        // it('constructs a DominantSpeakerSignaling with the DataTrackTransport,', () => {
+        //   assert(DominantSpeakerSignaling.calledWith(dataTrackTransport));
+        // });
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/room.js:1444]:
```suggestion
```

File: test/unit/spec/signaling/v2/room.js [hunk 8]
```diff
@@ -1622,9 +1482,9 @@ describe('RoomV2', () => {
             assert(dataTrackReceiver2.toDataTransport.calledOnce);
           });
 
-          it('constructs new NetworkQualitySignaling with the dataTrackReceiver2,', () => {
-            assert(DominantSpeakerSignaling.calledWith(dataTrackReceiver2));
-          });
+          // it('constructs new NetworkQualitySignaling with the dataTrackReceiver2,', () => {
+          //   assert(DominantSpeakerSignaling.calledWith(dataTrackReceiver2));
+          // });
```
Comment (manjeshbhargav) [inline test/unit/spec/signaling/v2/room.js:1487]:
```suggestion
```

**Review by manjeshbhargav — COMMENTED** [review]
@makarandp0 Getting better. I have a couple of questions:
1. Why was `updateSubscriberTrackPriority` and `setTrackPrioritySignaling` removed from `lib/signaling/participant.js`?
2. I see that there were mostly removal of unit tests and not new tests for the refactored code. Are you planning to add it in another commit?

**Comment by makarandp0** [comment]
Thanks @manjeshbhargav for taking a look I have incorporated your suggestions now.
> Why was updateSubscriberTrackPriority and setTrackPrioritySignaling removed from lib/signaling/participant.js?

`updateSubscriberTrackPriority` is moved to lib/signaling/v2/remoteparticipant.js because it's only applicable to RemoteParticipants. About `Participant!setTrackPrioritySignaling` - it was needed previously before because MSP signaling objects could get destroyed, and recreated for VMS failovers, With this refactor we are keeping Signaling class objects for the lifetime of room object, and just recreate transport objects inside those signaling objects. So  `Participant!setTrackPrioritySignaling` is not needed anymore. Any code to cache priorities is now moved to trackprioritysignaling.js

> I see that there were mostly removal of unit tests and not new tests for the refactored code. Are you planning to add it in another commit?

tests were removed because they are not applicable anymore (because room does not recreate signaling objects on VMS failovers). I see that we have same code coverage after refactored code  
https://38102-46595751-gh.circle-artifacts.com/0/coverage/lcov-report/lib/signaling/v2/mediasignaling.js.html
https://38102-46595751-gh.circle-artifacts.com/0/coverage/lcov-report/lib/signaling/v2/dominantspeakersignaling.js.html
https://38102-46595751-gh.circle-artifacts.com/0/coverage/lcov-report/lib/signaling/v2/networkqualitysignaling.js.html
But I noticed missing coverage for two branches and have added additional tests to validate refactored code in trackprioritysignaling.js

**Comment by makarandp0** [comment]
@manjeshbhargav, I have now added tests for `mediasignaling.js` - Please take another look

**Review by manjeshbhargav — APPROVED** [review]

**Pull request merged**